aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -218,36 +213,307 @@ Tensor NestedTensor_to_padded_tensor_cuda(
+    switch(backend){
+      case sdp::SDPBackend::flash_attention:
+          // TODO: enable flash attention kernel
+          return mem_efficient_helper_nested_unpacked(query_, key, value, dropout_p, need_attn_weights, is_causal);
+      case sdp::SDPBackend::efficient_attention:
+          return mem_efficient_helper_nested_unpacked(query_, key, value, dropout_p, need_attn_weights, is_causal);
+      case sdp::SDPBackend::math:
+        return at::_scaled_dot_product_attention_math(query_, key, value, attn_mask_, dropout_p, need_attn_weights, is_causal);
+      default:
+        TORCH_CHECK(false, "Unsupported backend for scaled_dot_product_attention");
+        return std::make_tuple(Tensor(), Tensor());
+    }
+}
+
+
+/**
+ * This function is used to calculate two pieces of metadata that are needed
+ * for use with flash-attention and efficient_attention kernels. They are the
+ * cumulative sequence_length over a batch of sequences and the maximum sequence
+ * length.
+ *
+ * @return A tuple of cumulative sequence lengths and the maximum sequence length
+ */
+std::tuple<Tensor, int64_t> cumulative_and_max_seq_len(Tensor qkv) {
+  TORCH_CHECK(
+      qkv.is_nested(),
+      "QKV must be nested for flash cumulative_seq_len calculation.")
+  auto* nt_impl = get_nested_tensor_impl(qkv);
+  const auto& sizes = nt_impl->get_nested_size_tensor();
+  auto size_tensor_stride = sizes.stride(0);
+
+  const int64_t batch_size = qkv.size(0);
+  auto cumulative_seqlen = at::zeros(
+      {batch_size + 1}, TensorOptions().device(at::kCPU).dtype(at::kInt));
+
+  auto* sizes_ptr = sizes.data_ptr<int64_t>();
+  auto* cumulative_seqlen_ptr = cumulative_seqlen.data_ptr<int32_t>();
+
+  int32_t sum = 0;
+  int64_t max_seqlen = -1;
+  cumulative_seqlen_ptr[0] = sum;
+  for (const auto i : c10::irange(batch_size)) {
+    // Calculate the cumulative sum of the sequence lengths
+    auto current_seq_len = sizes_ptr[(i * size_tensor_stride)];
+    sum += current_seq_len;
+    cumulative_seqlen_ptr[i + 1] = sum;
+
+    // Find the max element while we traverse
+    max_seqlen = std::max(max_seqlen, current_seq_len);
+  }
+  // Send to GPU, this is pretty light weight calc for normal batch size
+  // but maybe this needs to be on gpu
+  cumulative_seqlen = cumulative_seqlen.to(TensorOptions().device(at::kCUDA));
+  return std::tuple<Tensor, int64_t>{cumulative_seqlen, max_seqlen};
+}
+/**
+ * This function checks if the input query, key, value tensors are valid for
+ * use with the flash-attention and efficient_attention kernels without
+ * needing to call contiguous on the nested tensor input.
+ * It checks that the storage offsets' adjacent_difference constant over each of the
+ * tensor in the nested tensor and that the strides are monitonically decreasing.
+ * This is check is done after calling transpose on the nested tensor.
+ *
+ * @return A boolean indicating of contiguous needs to be called for input
+ */
+bool is_safe_to_get_storage_as_tensor(
```

Comment:
Why not do it for one NT at a time? I don't think you need to process all three at once.

Reply:
So I was wrong about the adjacent differences being constant throughout a tensor because the ragged dimension means the #of elements is going different as you jump through ntensor[i] to ntensor[i+1]. This currently checks that the adjacent differences between q,k,v though matched which would occur if q,k,v were chunked. I wonder if the only condition needed is consistent strides and that they are monotonically decreasing, pseudo contiguous. 

``` c++
q_storage_as_tensor.as_strided(
      {Nnz_q, num_heads, head_dim},
      {nnz_q_stride, head_q_stride, head_dim_stride},
      query_impl->get_storage_offsets()[0]);
```

nnz_q_stride, heqd_q_stride, head_dim_stride would be analogous to the "consistent. last dim" in this case. The ordering implies that the data is in fact laid out in batch, seq_len, num_heads, head_dim in memory.

The adjacent differences between tensors is trying to determine if the offset aspect is valid. We looking for the analogous "consistent offsets". I think the real condition would be that the adjacent difference for a singular nested_tensor is:

nt_offsets[i+1] - nt_offsets[i] = k * numel[i]

if this tensor is the result of chunk than k will be = #chunks